### PR TITLE
Add luacheck CI

### PR DESCRIPTION
**Pull request by theHamsta**
Thanks for this nice plugin! I was browsing a bit through the source code to see how it's working and I saw a few warnings.

I thought a style CI would be nice so I largely copied it from nvim-treesitter. The tests are failing at the moment (also on my local machine). This is why I activated `continue-on-error` for this step for now.

The style options are always very opinionated so you should review them on whether you agree on the policies (and changes) or not.

### CODE REVIEW

**Comment by theHamsta** [comment]
Actions run on my fork: https://github.com/theHamsta/neogit/actions

**Comment by TimUntersberger** [comment]
Thank you!

This was long overdue so I really appreciate you opening this PR. I will look at this later today or tomorrow.

File: lua/neogit/lib/util.lua [hunk 1]
```diff
@@ -124,7 +124,7 @@ return {
   filter = filter,
   str_right_pad = str_right_pad,
   str_count = str_count,
-  create_fold = create_fold,
+  --create_fold = create_fold, -- create_fold is not defined
```
Comment (theHamsta) [inline lua/neogit/lib/util.lua:127]: I couldn't find the definition of that function

**Comment by TimUntersberger** [comment]
Have you looked at [selene](https://github.com/Kampfkarren/selene) before? 

Selene looks more convenient to use than luacheck at first glance and it should be faster, because it is written in rust.

**Comment by theHamsta** [comment]
I didn't like selene. Especially in nvim-treesitter it didn't understand and luacheck had far more checks. I was really suprised why some would care for performance for something that runs in a few milliseconds. I mean luacheck is really fast you probably have to check a few 100 files to realize a difference

**Comment by TimUntersberger** [comment]
I'll trust your experience 👍

**Comment by theHamsta** [comment]
I'll add selene later. More additional linting can't harm.

**Comment by TimUntersberger** [comment]
@theHamsta I'd like to avoid having multiple linters. If anything is missing we should improve the linter we are using or switch to something else IMO

File: lua/neogit/lib/git/branch.lua [hunk 2]
```diff
@@ -1,4 +1,5 @@
 local a = require 'plenary.async_lib'
+--luacheck: ignore void
```
Comment (TimUntersberger) [inline lua/neogit/lib/git/branch.lua:2]: why is this needed?

File: lua/neogit/lib/git/status.lua [hunk 3]
```diff
@@ -3,24 +3,10 @@ local git = {
   stash = require("neogit.lib.git.stash")
 }
 local a = require 'plenary.async_lib'
-local async, await, await_all, future = a.async, a.await, a.await_all, a.future
+local async, await, await_all = a.async, a.await, a.await_all
 local util = require("neogit.lib.util")
 local Collection = require('neogit.lib.collection')
 
-local function marker_to_type(m)
```
Comment (TimUntersberger) [inline lua/neogit/lib/git/status.lua:10]: Why is this getting removed? Can we configure luacheck to tell us about unused functions, but not remove it?

File: lua/neogit/lib/md5.lua [hunk 4]
```diff
@@ -36,7 +36,9 @@ local bit_or, bit_and, bit_not, bit_xor, bit_rshift, bit_lshift
 
 local ok, bit = pcall(require, 'bit')
 if ok then
-  bit_or, bit_and, bit_not, bit_xor, bit_rshift, bit_lshift = bit.bor, bit.band, bit.bnot, bit.bxor, bit.rshift, bit.lshift
+  -- luacheck: ignore bit_not
```
Comment (TimUntersberger) [inline lua/neogit/lib/md5.lua:39]: why is this needed?

File: lua/neogit/lib/util.lua (hunk 1 above)
Comment (TimUntersberger) [reply lua/neogit/lib/util.lua:127]: Ah yeah you can remove this.

File: lua/neogit/popups/commit.lua [hunk 5]
```diff
@@ -163,7 +163,7 @@ local function create()
             -- we need \r? to support windows
             data = split(data, '\r?\n')
             await(prompt_commit_message(data, skip_gen))
-            local _, code = await(cli.commit.commit_message_file(COMMIT_FILE).args(unpack(popup.get_arguments())).call())
+            local _, code = await(cli.commit.commit_message_file(COMMIT_FILE).args(unpack(popup.get_arguments())).call()) --luacheck: ignore
```
Comment (TimUntersberger) [inline lua/neogit/popups/commit.lua:166]: why are you ignoring this line?

File: lua/neogit/status.lua [hunk 6]
```diff
@@ -5,7 +5,9 @@ local cli = require('neogit.lib.git.cli')
 local notif = require("neogit.lib.notification")
 local config = require("neogit.config")
 local a = require 'plenary.async_lib'
-local async, await, await_all, future, void, scheduler, run = a.async, a.await, a.await_all, a.future, a.void, a.scheduler, a.run
+-- luacheck: ignore run future
```
Comment (TimUntersberger) [inline lua/neogit/status.lua:8]: same thing here

File: tests/status_buf_spec.lua [hunk 7]
```diff
@@ -1,3 +1,5 @@
+-- Ignore line with only whitespace
+-- luacheck: ignore 611
```
Comment (TimUntersberger) [inline tests/status_buf_spec.lua:2]: I really don't like that we have to do this. I think we should disable this by default.

**Review by TimUntersberger — COMMENTED** [review]
Looks good overall, just have some questions regarding lint-customizations.

File: lua/neogit/lib/git/branch.lua (hunk 2 above)
Comment (theHamsta) [reply lua/neogit/lib/git/branch.lua:2]: Because "void" is unused. It can also just be removed

File: lua/neogit/lib/git/status.lua (hunk 3 above)
Comment (theHamsta) [reply lua/neogit/lib/git/status.lua:10]: You can ignore the warning. But better would be to just remove everything unused.

File: lua/neogit/popups/commit.lua (hunk 5 above)
Comment (theHamsta) [reply lua/neogit/popups/commit.lua:166]: Topo long. Line breaks look a bit look a bit weird but they are probably the better alternative

File: tests/status_buf_spec.lua (hunk 7 above)
Comment (theHamsta) [reply tests/status_buf_spec.lua:2]: I can also just remove the few remaining  lines with whites spaces. The warnings are all in the git commit messages

File: lua/neogit/status.lua (hunk 6 above)
Comment (theHamsta) [reply lua/neogit/status.lua]: Removed everything unused now. I originally thought that you may had a reason to always import this set of functions in the same manner in all files but they can just be imported when needed.

File: lua/neogit/lib/git/branch.lua (hunk 2 above)
Comment (TimUntersberger) [reply lua/neogit/lib/git/branch.lua]: Unused imports can be safely removed imo.

File: lua/neogit/lib/git/status.lua (hunk 3 above)
Comment (TimUntersberger) [reply lua/neogit/lib/git/status.lua:10]: 👍 to removing the function, I was just a bit concerned that some code might get removed that is accidently unused.

Comment (theHamsta) [reply lua/neogit/lib/git/status.lua:10]: -- luacheck: ignore marker_to_type
would silence the warning

**Comment by theHamsta** [comment]
One thought to selene: I think you can give it a try and migrate to it later if you like it better. But I wouldn't include it into this PR.

**Comment by TimUntersberger** [comment]
Thank you!
